odxtools/parameters/tablekeyparameter.py

```diff
@@ -133,46 +133,98 @@ def is_settable(self) -> bool:
         return True
 
     @override
-    def get_coded_value_as_bytes(self, encode_state: EncodeState) -> bytes:
-        tr_short_name = encode_state.parameter_values.get(self.short_name)
-
-        if tr_short_name is None:
-            # the table key has not been defined explicitly yet, but
-            # it is most likely implicitly defined by the associated
-            # TABLE-STRUCT parameters. Use all-zeros as a standin for
-            # the real data...
+    @final
+    def _encode_positioned_into_pdu(self, physical_value: Optional[ParameterValue],
+                                    encode_state: EncodeState) -> None:
+        # if you get this exception, you ought to use
+        # `.encode_placeholder_into_pdu()` followed by (after the
+        # value of the table key has been determined)
+        # `.encode_value_into_pdu()`.
+        raise RuntimeError("_encode_positioned_into_pdu() cannot be called for table keys.")
+
+    def encode_placeholder_into_pdu(self, physical_value: Optional[ParameterValue],
```

Comment:
Not possible to merge _encode_positioned_into_pdu and encode_placeholder_into_pdu into one?

Reply:
okay, I've just implemented a prototype doing it this way. It saves 10-20 lines of code, but I still don't like it because the semantics of `.encode_into_pdu()` would be completely different for non-key parameters compared to "normal" ones while exhibiting the same API (which would make bugs due to incorrect API usage very easy to do and hard to find). On the other hand, the proposed approach makes it quite obvious if `.encode_into_pdu()` is used incorrectly...